### PR TITLE
creating chef endpoints

### PR DESCRIPTION
**Pull request by eapg**
* Adding chef endpoints to access chef controller from navigator

### CODE REVIEW

File: src/flask/chef_app.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+class ChefSchema(ma.Schema):
+    class Meta:
+        fields = ("id", "user_id", "skill")
+
+
+chef_schema = ChefSchema()
+chefs_schema = ChefSchema(many=True)
+
+
+@app.route("/chefs", methods=["POST"])
```
Comment (reynoldmorel) [inline src/flask/chef_app.py:27]: You should put this in the controller

File: src/flask/chef_app.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+app = Flask(__name__)
+
+ma = Marshmallow(app)
+
+ioc = Injector(DiProviders)
+chef_controller = ioc.get(ChefController)
+
+
+# schemas
+class ChefSchema(ma.Schema):
```
Comment (reynoldmorel) [inline src/flask/chef_app.py:18]: Create a file for Schemas

File: src/flask/chef_app.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+from src.api.controllers.chef_controller import ChefController
+from src.constants.audit import Status, InternalUsers
+from src.core.di_config import DiProviders
+from src.lib.entities.sqlalchemy_orm_mapping import Chef
+
+app = Flask(__name__)
+
+ma = Marshmallow(app)
+
+ioc = Injector(DiProviders)
```
Comment (reynoldmorel) [inline src/flask/chef_app.py:13]: I think flask already includes an implementation for dependency injector

File: src/flask/chef_app.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+
+@app.route("/chefs", methods=["POST"])
+def create_chef():
+
+    new_chef = Chef()
+    new_chef.user_id = request.json["user_id"]
+    new_chef.skill = request.json["skill"]
+    new_chef.entity_status = Status.ACTIVE.value
+    new_chef.created_by = InternalUsers.SEEDER.value
```
Comment (reynoldmorel) [inline src/flask/chef_app.py:34]: This is not the seeder, create a new user for this one, maybe an API_USER

File: src/flask/chef_app.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+@app.route("/chefs/<chef_id>", methods=["DELETE"])
+def delete_chef_by_id(chef_id):
+    chef = Chef()
+    chef.updated_by = InternalUsers.SEEDER.value
+    chef_controller.delete_by_id(chef_id, chef)
+    return "Chef Deleted"
+
+
+if __name__ == "__main__":
+    app.run(debug=True)
```
Comment (reynoldmorel) [inline src/flask/chef_app.py:71]: You should setup dependency injector here

Comment (reynoldmorel) [inline src/flask/chef_app.py:71]: You should setup dependency injector here

**Review by reynoldmorel — CHANGES_REQUESTED** [review]

File: src/flask/schemas/chef_schema.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+    user_id = fields.Int()
+    skill = fields.Int()
+
+    @post_load
+    def create_sqlalchemy_chef_entity(self, data, **kwargs):
+        chef = Chef()
+        chef.skill = data["skill"]
+        chef.user_id = data["user_id"]
+        chef.entity_status = Status.ACTIVE.value
```
Comment (reynoldmorel) [inline src/flask/schemas/chef_schema.py:20]: You don't need this, your service should this for you. Also use `data.get()` instead. It is possible that a key may not exist. You can create a generic utility that makes this conversion for you by getting all the keys from and set all the attributes of the class having the same props

File: src/flask/routes/chef.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+chef_controller = ioc.get(ChefController)
+chef_schema = ChefSchema()
+chef_schemas = ChefSchema(many=True)
+
+
+@chef_blueprint.route("/chefs", methods=["POST"])
+def add():
+    json_chef_data = request.get_json()
+    chef = chef_schema.load(json_chef_data)
+    chef_controller.add(chef)
```
Comment (reynoldmorel) [inline src/flask/routes/chef.py:24]: Return the value from the controller instead, stop relying on the object references. A references can change at any moment.

File: src/flask/routes/chef.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+from flask import Blueprint, request
+from injector import Injector
+
+from src.api.controllers.chef_controller import ChefController
+from src.constants.audit import ExternalUsers
+from src.core.di_config import DiProviders
+from src.flask.schemas.chef_schema import ChefSchema
+from src.lib.entities.sqlalchemy_orm_mapping import Chef
+
+chef_blueprint = Blueprint("chef_routes", __name__)
```
Comment (reynoldmorel) [inline src/flask/routes/chef.py:12]: Could you wrap everything into a setup function like this:

```
def setup_chef_routes(ioc, ):
   chef_blueprint = ....
   chef_schema = ....

  chef_blueprint.route(...)
  def add...

 return {'chef_blueprint': chef_blueprint, ....}
```

That way we can reuse ioc instance and any other element

File: src/flask/routes/chef.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+def get_all():
+    chefs = chef_controller.get_all()
+    return chef_schemas.dump(chefs)
+
+
+@chef_blueprint.route("/chefs/<chef_id>", methods=["PUT"])
+def update_chef_by_id(chef_id):
+    json_chef_data = request.get_json()
+    chef = chef_schema.load(json_chef_data)
+    chef.updated_by = ExternalUsers.API_USER.value
```
Comment (reynoldmorel) [inline src/flask/routes/chef.py:45]: Set the user through the client `postman` for now, do not harcdode it

File: src/flask/run_flask_app.py [hunk 10]
```diff
@@ -0,0 +1,13 @@
+from flask_injector import FlaskInjector
+
+from flask import Flask
+from src.core.di_config import DiProviders
+from src.flask.routes.chef import chef_blueprint
+
+app = Flask(__name__)
+app.register_blueprint(chef_blueprint)
```
Comment (reynoldmorel) [inline src/flask/run_flask_app.py:8]: Wrap this into a setup function as well.

```
def setup _api(app):
```

File: src/flask/run_flask_app.py [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+
+FlaskInjector(app=app, modules=[DiProviders])
+
+if __name__ == "__main__":
+    app.run(debug=True)
```
Comment (reynoldmorel) [inline src/flask/run_flask_app.py:13]: Call the setup here

File: src/flask/run_flask_app.py (hunk 10 above)
Comment (reynoldmorel) [reply src/flask/run_flask_app.py:8]: Call chef setup function here and get the ioc here to be send the chef setup function

File: src/flask/schemas/chef_schema.py [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+        chef.created_by = InternalUsers.SEEDER.value
```
Comment (reynoldmorel) [inline src/flask/schemas/chef_schema.py:21]: As mentioned above send this from the client

**Review by reynoldmorel — CHANGES_REQUESTED** [review]

File: src/flask/routes/chef.py (hunk 8 above)
Comment (eapg) [reply src/flask/routes/chef.py:12]: done.

File: src/flask/run_flask_app.py (hunk 10 above)
Comment (eapg) [reply src/flask/run_flask_app.py:8]: done.

File: src/flask/schemas/chef_schema.py (hunk 12 above)
Comment (eapg) [reply src/flask/schemas/chef_schema.py:21]: done.

File: src/flask/run_flask_app.py (hunk 11 above)
Comment (eapg) [reply src/flask/run_flask_app.py:13]: done.

File: src/flask/routes/chef.py (hunk 9 above)
Comment (eapg) [reply src/flask/routes/chef.py:45]: done.

File: src/flask/routes/chef.py (hunk 7 above)
Comment (eapg) [reply src/flask/routes/chef.py:24]: done.

File: src/flask/schemas/chef_schema.py (hunk 6 above)
Comment (eapg) [reply src/flask/schemas/chef_schema.py:20]: done. I used data.get() instead.

File: src/flask/api_setup.py [hunk 13]
```diff
@@ -0,0 +1,15 @@
+from injector import Injector
+
+from src.core.di_config import DiProviders
+from src.flask.routes.chef import setup_chef_routes
+
+
+def setup_api(app):
```
Comment (reynoldmorel) [inline src/flask/api_setup.py:7]: Return the app

File: src/flask/run_flask_app.py [hunk 14]
```diff
@@ -0,0 +1,10 @@
+from flask import Flask
+
+from src.flask.api_setup import setup_api
+
+app = Flask(__name__)
```
Comment (reynoldmorel) [inline src/flask/run_flask_app.py:5]: Rename to `flaskApp`

File: src/flask/run_flask_app.py [hunk 15]
```diff
@@ -0,0 +1,10 @@
...
+
+
+if __name__ == "__main__":
+    setup_api(app)
```
Comment (reynoldmorel) [inline src/flask/run_flask_app.py:9]: Change to

```
app = setup_api(flaskApp)
```

File: src/flask/schemas/chef_schema.py [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+class ChefSchema(Schema):
+
+    user_id = fields.Int()
+    skill = fields.Int()
+    created_by = fields.Int()
+    updated_by = fields.Int()
+
+    @post_load
+    def create_sqlalchemy_chef_entity(self, data, **kwargs):
+        chef = Chef()
```
Comment (reynoldmorel) [inline src/flask/schemas/chef_schema.py:15]: What about creating the general utility for this?

**Review by reynoldmorel — CHANGES_REQUESTED** [review]

File: src/flask/api_setup.py (hunk 13 above)
Comment (eapg) [reply src/flask/api_setup.py:7]: done.

File: src/flask/run_flask_app.py (hunk 14 above)
Comment (eapg) [reply src/flask/run_flask_app.py:5]: done

File: src/flask/schemas/chef_schema.py (hunk 16 above)
Comment (eapg) [reply src/flask/schemas/chef_schema.py:15]: done.

**Review by reynoldmorel — APPROVED** [review]

**Review by reynoldmorel — CHANGES_REQUESTED** [review]
Add tests

File: src/constants/audit.py [hunk 17]
```diff
@@ -10,3 +10,7 @@ class InternalUsers(Enum):
     SEEDER = 1
     KITCHEN_SIMULATOR = 2
     ETL = 3
+
+
+class ExternalUsers(Enum):
+    API_USER = 4
```
Comment (reynoldmorel) [inline src/constants/audit.py:16]: Move this to internal users class

File: src/tests/flask/base_flask_setup_test.py [hunk 18]
```diff
@@ -0,0 +1,33 @@
+from abc import abstractmethod
+
+from flask import Flask
+from src.core.sqlalchemy_config import get_engine
+from src.flask.api_setup import setup_api
+from src.lib.entities.sqlalchemy_orm_mapping import Base
+from src.tests.base_env_config_test import BaseEnvConfigTest
+
+engine = get_engine()
```
Comment (reynoldmorel) [inline src/tests/flask/base_flask_setup_test.py:9]: Do this in the setup method. No need to do this outside of the class

File: src/tests/flask/base_flask_setup_test.py [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+
+
+def clean_database():
```
Comment (reynoldmorel) [inline src/tests/flask/base_flask_setup_test.py:12]: Add engine as arg of this func

File: src/tests/flask/routes/chef_routes_test.py [hunk 20]
```diff
@@ -0,0 +1,39 @@
+from src.tests.flask.base_flask_setup_test import BaseFlaskSetupTest
+
+
+class ChefApiTest(BaseFlaskSetupTest):
```
Comment (reynoldmorel) [inline src/tests/flask/routes/chef_routes_test.py:4]: Rename to `ChefApiE2ETest` and the file as well `chef_api_e2e_test.py`.

File: src/tests/flask/utils/api_util_test.py [hunk 21]
```diff
@@ -0,0 +1,16 @@
+from src.flask.utils.api_util import fetch_class_from_data
+import unittest
+
+from src.lib.entities.chef import Chef
+
+
+class FetchClassFromDataTest(unittest.TestCase):
```
Comment (reynoldmorel) [inline src/tests/flask/utils/api_util_test.py:7]: Did you mean `ApiUtilTest`?

**Review by reynoldmorel — CHANGES_REQUESTED** [review]
Create a folder `tests/e2e/api` and place your tests there, no need to specify flask since this is a client.

Add tests for remaining endpoints in the chef controller

File: src/tests/flask/routes/chef_routes_test.py [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+    def after_base_setup(self):
+        self.json_chef = {"user_id": 1, "skill": 5, "created_by": 1, "updated_by": 1}
```
[the remark this answers is not in the document]
Comment (eapg) [reply src/tests/flask/routes/chef_routes_test.py:6]: done.

File: src/constants/audit.py (hunk 17 above)
Comment (eapg) [reply src/constants/audit.py:16]: done.

File: src/tests/flask/base_flask_setup_test.py (hunk 18 above)
Comment (eapg) [reply src/tests/flask/base_flask_setup_test.py:9]: done.

File: src/tests/flask/base_flask_setup_test.py (hunk 19 above)
Comment (eapg) [reply src/tests/flask/base_flask_setup_test.py:12]: done.

File: src/tests/flask/routes/chef_routes_test.py (hunk 20 above)
Comment (eapg) [reply src/tests/flask/routes/chef_routes_test.py:4]: done.

File: src/tests/flask/utils/api_util_test.py (hunk 21 above)
Comment (eapg) [reply src/tests/flask/utils/api_util_test.py:7]: done.

File: src/tests/flask/routes/chef_api_e2e_test.py [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+        request_updated_parameter = request.get_json()
+        self.assertEqual(10, request_updated_parameter["skill"])
+
+    def test_deleted_chef_request_successfully(self):
+        self.client.post("/chefs", json=self.json_chef)
+        request = self.client.delete("/chefs/1", json={"updated_by": 3})
+        self.assertEqual(request.get_data(as_text=True), "Chef 1 Deleted")
+
+    def test_get_available_chef_request_successfully(self):
+        with self.engine.begin() as conn:
```
Comment (reynoldmorel) [inline src/tests/flask/routes/chef_api_e2e_test.py:49]: Move this to the order fixture, and please pass down arguments to set the different props, in this case status

File: src/tests/flask/routes/chef_api_e2e_test.py [hunk 24]
```diff
@@ -0,0 +1,55 @@
...
+            conn.execute(text("CALL insert_order_with_defaults(order_status:='IN_PROCESS');"))
+        self.client.post("/chefs", json=self.json_chef)
```
Comment (reynoldmorel) [inline src/tests/flask/routes/chef_api_e2e_test.py:51]: Compare again created users

File: src/tests/flask/routes/chef_api_e2e_test.py [hunk 25]
```diff
@@ -0,0 +1,55 @@
+from sqlalchemy import text
+
+from src.tests.flask.base_flask_setup_test import BaseFlaskSetupTest
+from src.tests.utils.fixtures.json_entities_fixture import build_chef
+
+QUERY_TO_INSERT_AN_ORDER = """
```
Comment (reynoldmorel) [inline src/tests/flask/routes/chef_api_e2e_test.py:6]: Do you need this?

**Review by reynoldmorel — CHANGES_REQUESTED** [review]

File: src/tests/flask/routes/chef_api_e2e_test.py (hunk 25 above)
Comment (eapg) [reply src/tests/flask/routes/chef_api_e2e_test.py:6]: done.

File: src/tests/flask/routes/chef_api_e2e_test.py (hunk 24 above)
Comment (eapg) [reply src/tests/flask/routes/chef_api_e2e_test.py:51]: done.

File: src/tests/flask/routes/chef_api_e2e_test.py (hunk 23 above)
Comment (eapg) [reply src/tests/flask/routes/chef_api_e2e_test.py:49]: done.

File: src/tests/e2e/api/chef_api_e2e_test.py [hunk 26]
```diff
@@ -0,0 +1,61 @@
...
+    def test_deleted_chef_request_successfully(self):
+        json_chef = build_chef()
+        self.client.post("/chefs", json=json_chef)
+        request = self.client.delete("/chefs/1", json={"updated_by": 3})
+        self.assertEqual(request.get_data(as_text=True), "Chef 1 Deleted")
+
+    def test_get_available_chef_request_successfully(self):
+        json_chef_1 = build_chef(chef_id=1)
+        json_chef_2 = build_chef(chef_id=2)
+        create_new_order_with_procedure(engine=self.engine, order_status="IN_PROCESS")
```
Comment (reynoldmorel) [inline src/tests/e2e/api/chef_api_e2e_test.py:56]: Rename to `cereate_order_with_procedure`

Comment (reynoldmorel) [reply src/tests/e2e/api/chef_api_e2e_test.py:56]: Could you send the chef id you want to use?

File: src/tests/e2e/api/chef_api_e2e_test.py [hunk 27]
```diff
@@ -0,0 +1,61 @@
...
+        chef_1 = self.client.post("/chefs", json=json_chef_1)
+        chef_2 = self.client.post("/chefs", json=json_chef_2)
+        request = self.client.get("/chefs/available")
```
Comment (reynoldmorel) [inline src/tests/e2e/api/chef_api_e2e_test.py:59]: This should return a list of chef objects, the client will not do anything with an id. That means the client would need to do more calls to the api to get these chefs ids shef information and also make more code on his end to map them.

**Review by reynoldmorel — CHANGES_REQUESTED** [review]

File: src/tests/e2e/api/chef_api_e2e_test.py (hunk 27 above)
Comment (eapg) [reply src/tests/e2e/api/chef_api_e2e_test.py:59]: done.

File: src/tests/e2e/api/chef_api_e2e_test.py (hunk 26 above)
Comment (eapg) [reply src/tests/e2e/api/chef_api_e2e_test.py:56]: done.

**Review by reynoldmorel — APPROVED** [review]

**Pull request merged**